File: test-utils/src/main/java/ru/virgil/test_utils/AssertUtils.java
package ru.virgil.test_utils;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.truth.Truth;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.StreamSupport;

/**
 * Помогает делать всякие сложные ассерты.
 */
@Component
@RequiredArgsConstructor
public class AssertUtils {

    private final ObjectMapper objectMapper;

    public void partialEquals(Object full, Object partial) {
        List<JsonNode> fullFields = StreamSupport.stream(objectMapper.valueToTree(full).spliterator(), false)
                .toList();
        List<JsonNode> partialFields = StreamSupport.stream(objectMapper.valueToTree(partial).spliterator(), false)
                .filter(jsonNode -> !jsonNode.isNull())
                .toList();
        Truth.assertThat(fullFields).containsAtLeastElementsIn(partialFields);
    }

}

File: test-utils/src/main/java/ru/virgil/test_utils/AssertUtils.kt
package ru.virgil.test_utils

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.common.truth.Truth
import org.springframework.stereotype.Component

/**
 * Помогает делать всякие сложные ассерты.
 */
@Component
class AssertUtils(private val objectMapper: ObjectMapper) {

    fun partialEquals(full: Any, partial: Any) {
        val fullJsonNode = objectMapper.valueToTree<JsonNode>(full)
        val fullFields = fullJsonNode.toList()
        val partialJsonNode = objectMapper.valueToTree<JsonNode>(partial)
        val partialFields = partialJsonNode.filterNotNull()
            .filter { !it.isNull }
            .toList()
        Truth.assertThat(fullFields).containsAtLeastElementsIn(partialFields)
    }
}
